lib/diagnostics_channel.js

```diff
@@ -0,0 +1,113 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  ObjectCreate,
+  ObjectSetPrototypeOf
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const { triggerUncaughtException } = internalBinding('errors');
+
+const { WeakReference } = internalBinding('util');
+
+// TODO(qard): should there be a C++ channel interface?
+class ActiveChannel {
+  subscribe(subscription) {
+    if (typeof subscription !== 'function') {
+      throw new ERR_INVALID_ARG_TYPE('subscription', ['function'],
+                                     subscription);
+    }
+    ArrayPrototypePush(this._subscribers, subscription);
+  }
+
+  unsubscribe(subscription) {
+    const index = ArrayPrototypeIndexOf(this._subscribers, subscription);
+    if (index >= 0) {
+      ArrayPrototypeSplice(this._subscribers, index, 1);
+
+      // When there are no more active subscribers, restore to fast prototype.
+      if (!this._subscribers.length) {
+        // eslint-disable-next-line no-use-before-define
+        ObjectSetPrototypeOf(this, Channel.prototype);
+      }
```

Comment:
this seems like very strange observable behavior for any class to have. Why is this necessary?

Reply:
It's orders of magnitude faster this way. The two prototypes are externally identical, changing from doing nothing at all when there's no subscribers.

Reply:
There's no way to do that in the method implementations instead of by swapping the entire prototype?

Reply:
Doing it in the methods is many, many times slower. It's roughly 20-40x faster by swapping the prototypes. This API is meant to be a data firehose, so it's absolutely _critical_ that these methods are as fast as possible, which means that they should ideally be doing _literally nothing_ unless there's a reason for them to do otherwise.

Reply:
I've used `Symbol.species` to ensure it always passes `channel instanceof Channel`.

Reply:
I've added a `Symbol.hasInstance` method so it always passes `instanceof` checks.